File: android/app/src/main/java/dev/wispers/access/android/proxy/HeaderRewriter.kt
package dev.wispers.access.android.proxy

import dev.wispers.access.android.storage.ShareId

/** A single HTTP header. Mutable so rewriters can edit in place. */
data class Header(var name: String, var value: String)

/**
 * Pluggable header transforms applied at the proxy boundary, both directions.
 *
 * Operates in place on an ordered, duplicate-allowing list (Set-Cookie can appear
 * multiple times in a response). The proxy core strips hop-by-hop headers and
 * manages framing (Content-Length / Transfer-Encoding) around the rewriter call,
 * so a rewriter can't accidentally break framing.
 */
interface HeaderRewriter {
    fun rewriteRequest(
        method: String,
        target: String,
        headers: MutableList<Header>,
        shareId: ShareId,
    )

    fun rewriteResponse(
        status: Int,
        headers: MutableList<Header>,
        shareId: ShareId,
    )
}

class DefaultHeaderRewriter : HeaderRewriter {
    override fun rewriteRequest(
        method: String,
        target: String,
        headers: MutableList<Header>,
        shareId: ShareId,
    ) {
        // No-op for now. Hooks for Host rewriting, Origin/Referer scrubbing land here.
    }

    override fun rewriteResponse(
        status: Int,
        headers: MutableList<Header>,
        shareId: ShareId,
    ) {
        // Strip Set-Cookie Domain= attribute so cookies bind to <shareId>.localhost
        // (the host the WebView talks to), not whatever the upstream thought it was.
        for (h in headers) {
            if (h.name.equals("set-cookie", ignoreCase = true)) {
                h.value = stripDomainAttribute(h.value)
            }
        }
    }

    private fun stripDomainAttribute(setCookie: String): String =
        setCookie.split(";")
            .filterNot { it.trim().startsWith("domain=", ignoreCase = true) }
            .joinToString(";")
}

File: ios/Wispers Access/Tunnel/HeaderRewriter.swift
import Foundation

/// A single HTTP header. Mutable so a rewriter can edit it in place.
struct HTTPHeader: Sendable {
    var name: String
    var value: String
}

/// Header transforms applied at the proxy boundary, both directions. The proxy
/// core manages framing and hop-by-hop headers around these calls, so a rewrite
/// can't accidentally break framing. Port of the Android `HeaderRewriter`.
nonisolated struct HeaderRewriter: Sendable {
    func rewriteRequest(
        method: String,
        target: String,
        headers: inout [HTTPHeader],
        shareID: ShareID
    ) {
        // The proxy-auth cookie authenticates the web view to the loopback
        // proxy; it must not travel upstream.
        for i in headers.indices where headers[i].name.lowercased() == "cookie" {
            headers[i].value = Self.stripProxyAuthPair(headers[i].value)
        }
        headers.removeAll { $0.name.lowercased() == "cookie" && $0.value.isEmpty }
    }

    func rewriteResponse(
        status: Int,
        headers: inout [HTTPHeader],
        shareID: ShareID
    ) {
        // The upstream app mustn't clobber the proxy-auth cookie in the web
        // view's store.
        headers.removeAll {
            $0.name.lowercased() == "set-cookie"
                && ProxyAuth.cookiePairs($0.value).first?.name == ProxyAuth.cookieName
        }
        // Strip the Set-Cookie `Domain=` attribute so cookies bind to the
        // loopback origin the WebView actually talks to, not whatever host the
        // upstream app assumed it was running on.
        for i in headers.indices where headers[i].name.lowercased() == "set-cookie" {
            headers[i].value = Self.stripDomainAttribute(headers[i].value)
        }
    }

    private static func stripProxyAuthPair(_ cookie: String) -> String {
        cookie
            .split(separator: ";")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.hasPrefix(ProxyAuth.cookieName + "=") }
            .joined(separator: "; ")
    }

    private static func stripDomainAttribute(_ setCookie: String) -> String {
        setCookie
            .split(separator: ";", omittingEmptySubsequences: false)
            .filter { !$0.trimmingCharacters(in: .whitespaces).lowercased().hasPrefix("domain=") }
            .joined(separator: ";")
    }
}
